### PR TITLE
Feat: Port All Delta-V Mail Improvements

### PR DESCRIPTION
**Pull request by Mnemotechnician**
# Description
Showing some love to our hard-working couriers.

This cherry-picks the following PRs from delta-v:
- https://github.com/DeltaV-Station/Delta-v/pull/1472 - Adds a MailMetrics PDA cartridge for couriers and the LO
- https://github.com/DeltaV-Station/Delta-v/pull/1652 - Ports mail tweaks from frontier to delta-v, adding large packages, RPDs (mail cannons), and more mail
- https://github.com/DeltaV-Station/Delta-v/pull/1788 - Adding a couple more packages
- https://github.com/DeltaV-Station/Delta-v/pull/1925 - Fixing the non-functional "last known location" part of mail descriptions

Some mail items, such as the opporozidone syringe and rainbow joints/blunts, had to be disabled because we don't have them (yet?) 

<details><summary><h1>Media</h1></summary>
<p>

image

image

image

image

image

image

</p>
</details>

---

# Changelog
:cl:
- add: The Courier and Logistics Officer now have a new program in their PDA for tracking mail delivery performance, including earnings and percent of packages opened, damaged, or expired.
- add: The list of possible mail packages has been greately expanded, and now includes large parcels.
- add: The CourierDrobe now offers a rapid mail delivery device, along with capsules for it.

### CODE REVIEW

**Review by VMSolidus — CHANGES_REQUESTED** [review]
There's simply too many to count here, but everything in this PR needs to be moved out of specific fork namespaces, and down into the core namespace.

**Comment by Mnemotechnician** [comment]
> There's simply too many to count here, but everything in this PR needs to be moved out of specific fork namespaces, and down into the core namespace.

ugh, remind me like on Saturday maybe...

File: Content.Server/Nyanotrasen/Mail/Components/MailTeleporterComponent.cs [hunk 1]
```diff
@@ -102,7 +102,21 @@ public sealed partial class MailTeleporterComponent : Component
...
+        /// </summary>
+        [DataField]
+        public int LargeBonus = 1500; //DeltaV; 5000 to 1500
+
+        /// <summary>
+        /// What's the malus for failing to deliver a large package?
+        /// </summary>
+        [DataField]
+        public int LargeMalus = -500; //DeltaV; -250 to -500
+        // End Frontier: Large mail
```
Comment (VMSolidus) [inline Content.Server/Nyanotrasen/Mail/Components/MailTeleporterComponent.cs:120]:
```suggestion
        /// <summary>
        /// What's the bonus for delivering a large package intact?
        /// </summary>
        [DataField]
        public int LargeBonus = 1500; //DeltaV; 5000 to 1500

        /// <summary>
        /// What's the malus for failing to deliver a large package?
        /// </summary>
        [DataField]
        public int LargeMalus = -500; //DeltaV; -250 to -500
```

File: Content.Server/Nyanotrasen/Mail/Components/MailTeleporterComponent.cs [hunk 2]
```diff
@@ -27,8 +27,8 @@ public sealed partial class MailTeleporterComponent : Component
         /// The MailDeliveryPoolPrototype that's used to select what mail this
         /// teleporter can deliver.
         /// </summary>
-        [DataField("mailPool")]
-        public string MailPool = "RandomMailDeliveryPool";
+        [DataField]
+        public string MailPool = "RandomDeltaVMailDeliveryPool"; // Frontier / DeltaV: Mail rework
```
Comment (VMSolidus) [inline Content.Server/Nyanotrasen/Mail/Components/MailTeleporterComponent.cs:31]:
```suggestion
        public string MailPool = "RandomDeltaVMailDeliveryPool";
```

File: Content.Server/Nyanotrasen/Mail/Components/MailComponent.cs [hunk 3]
```diff
@@ -58,43 +55,51 @@ public sealed partial class MailComponent : SharedMailComponent
...
+        [DataField]
         public bool IsPriority = false;
 
+        // Frontier Mail Port: large mail
+        /// <summary>
+        /// Whether this parcel is large.
+        /// </summary>
+        [DataField]
+        public bool IsLarge = false;
+        // End Frontier: large mail
```
Comment (VMSolidus) [inline Content.Server/Nyanotrasen/Mail/Components/MailComponent.cs:67]:
```suggestion
        /// <summary>
        ///     Whether this parcel is large.
        /// </summary>
        [DataField]
        public bool IsLarge;
```

File: Content.Server/Nyanotrasen/Mail/MailCommands.cs [hunk 4]
```diff
@@ -56,6 +57,16 @@ public async void Execute(IConsoleShell shell, string argStr, string[] args)
...
 
+        // Frontier: Large Mail
+        bool isLarge = false;
+        if (args.Length > 4 && !Boolean.TryParse(args[4], out isLarge))
+        {
+            shell.WriteError(Loc.GetString("shell-invalid-bool"));
+            return;
+        }
+        var mailPrototype = isLarge ? _blankLargeMailPrototype : _blankMailPrototype;
+        // End Frontier
```
Comment (VMSolidus) [inline Content.Server/Nyanotrasen/Mail/MailCommands.cs:68]:
```suggestion
        var isLarge = false;
        if (args.Length > 4 && !Boolean.TryParse(args[4], out isLarge))
        {
            shell.WriteError(Loc.GetString("shell-invalid-bool"));
            return;
        }
        var mailPrototype = isLarge ? _blankLargeMailPrototype : _blankMailPrototype;
```

File: Content.Server/Nyanotrasen/Mail/MailCommands.cs [hunk 5]
```diff
@@ -66,9 +77,9 @@ public async void Execute(IConsoleShell shell, string argStr, string[] args)
             return;
         }
 
-        if (!_prototypeManager.HasIndex<EntityPrototype>(_blankMailPrototype))
+        if (!_prototypeManager.HasIndex<EntityPrototype>(mailPrototype)) // Frontier: _blankMailPrototype<mailPrototype
```
Comment (VMSolidus) [inline Content.Server/Nyanotrasen/Mail/MailCommands.cs:80]:
```suggestion
        if (!_prototypeManager.HasIndex<EntityPrototype>(mailPrototype))
```

File: Content.Server/Nyanotrasen/Mail/MailCommands.cs [hunk 6]
```diff
@@ -66,9 +77,9 @@ public async void Execute(IConsoleShell shell, string argStr, string[] args)
...
         {
-            shell.WriteLine(Loc.GetString("command-mailto-no-blankmail", ("blankMail", _blankMailPrototype)));
+        shell.WriteLine(Loc.GetString("command-mailto-no-blankmail", ("blankMail", mailPrototype))); // Frontier: _blankMailPrototype<mailPrototype
```
Comment (VMSolidus) [inline Content.Server/Nyanotrasen/Mail/MailCommands.cs:82]:
```suggestion
        shell.WriteLine(Loc.GetString("command-mailto-no-blankmail", ("blankMail", mailPrototype)));
```

File: Content.Server/Nyanotrasen/Mail/MailCommands.cs [hunk 7]
```diff
@@ -90,12 +101,12 @@ public async void Execute(IConsoleShell shell, string argStr, string[] args)
             return;
         }
 
-        var mailUid = _entityManager.SpawnEntity(_blankMailPrototype, _entityManager.GetComponent<TransformComponent>(containerUid).Coordinates);
+        var mailUid = _entityManager.SpawnEntity(mailPrototype, _entityManager.GetComponent<TransformComponent>(containerUid).Coordinates); // Frontier: _blankMailPrototype<mailPrototype
```
Comment (VMSolidus) [inline Content.Server/Nyanotrasen/Mail/MailCommands.cs:104]:
```suggestion
        var mailUid = _entityManager.SpawnEntity(mailPrototype, _entityManager.GetComponent<TransformComponent>(containerUid).Coordinates);
```

File: Content.Server/Nyanotrasen/Mail/MailCommands.cs [hunk 8]
```diff
@@ -90,12 +101,12 @@ public async void Execute(IConsoleShell shell, string argStr, string[] args)
...
         var mailContents = _containerSystem.EnsureContainer<Container>(mailUid, _mailContainer);
 
         if (!_entityManager.TryGetComponent(mailUid, out MailComponent? mailComponent))
         {
-            shell.WriteLine(Loc.GetString("command-mailto-bogus-mail", ("blankMail", _blankMailPrototype), ("requiredMailComponent", nameof(MailComponent))));
+            shell.WriteLine(Loc.GetString("command-mailto-bogus-mail", ("blankMail", mailPrototype), ("requiredMailComponent", nameof(MailComponent)))); // Frontier: _blankMailPrototype<mailPrototype
```
Comment (VMSolidus) [inline Content.Server/Nyanotrasen/Mail/MailCommands.cs:109]:
```suggestion
            shell.WriteLine(Loc.GetString("command-mailto-bogus-mail", ("blankMail", mailPrototype), ("requiredMailComponent", nameof(MailComponent))));
```

File: Content.Server/Nyanotrasen/Mail/MailCommands.cs [hunk 9]
```diff
@@ -104,6 +115,7 @@ public async void Execute(IConsoleShell shell, string argStr, string[] args)
 
         mailComponent.IsFragile = isFragile;
         mailComponent.IsPriority = isPriority;
+        mailComponent.IsLarge = isLarge; //Frontier Mail
```
Comment (VMSolidus) [inline Content.Server/Nyanotrasen/Mail/MailCommands.cs:118]:
```suggestion
        mailComponent.IsLarge = isLarge;
```

File: Content.Server/Nyanotrasen/Mail/MailSystem.cs [hunk 10]
```diff
@@ -78,6 +80,9 @@ public sealed class MailSystem : EntitySystem
         [Dependency] private readonly MindSystem _mindSystem = default!;
         [Dependency] private readonly MetaDataSystem _metaDataSystem = default!;
 
+        // DeltaV - system that keeps track of mail and cargo stats
```
Comment (VMSolidus) [inline Content.Server/Nyanotrasen/Mail/MailSystem.cs:83]:
```suggestion
```

File: Content.Server/Nyanotrasen/Mail/MailSystem.cs [hunk 11]
```diff
@@ -222,6 +227,13 @@ private void OnAfterInteractUsing(EntityUid uid, MailComponent component, AfterI
                 }
             }
 
+            // DeltaV - Add earnings to logistic stats
```
Comment (VMSolidus) [inline Content.Server/Nyanotrasen/Mail/MailSystem.cs:230]:
```suggestion
```

File: Content.Server/Nyanotrasen/Mail/MailSystem.cs [hunk 12]
```diff
@@ -241,19 +253,19 @@ private void OnAfterInteractUsing(EntityUid uid, MailComponent component, AfterI
                     continue;
 
                 _cargoSystem.UpdateBankAccount(station, account, component.Bounty);
-                return;
             }
         }
 
         private void OnExamined(EntityUid uid, MailComponent component, ExaminedEvent args)
         {
+            MailEntityStrings mailEntityStrings = component.IsLarge ? MailConstants.MailLarge : MailConstants.Mail; //Frontier: mail types stored per type (large mail)
```
Comment (VMSolidus) [inline Content.Server/Nyanotrasen/Mail/MailSystem.cs:261]:
```suggestion
            MailEntityStrings mailEntityStrings = component.IsLarge ? MailConstants.MailLarge : MailConstants.Mail;
```

File: Content.Server/Nyanotrasen/Mail/MailSystem.cs [hunk 13]
```diff
@@ -241,19 +253,19 @@ private void OnAfterInteractUsing(EntityUid uid, MailComponent component, AfterI
...
             if (!args.IsInDetailsRange)
             {
-                args.PushMarkup(Loc.GetString("mail-desc-far"));
+                args.PushMarkup(Loc.GetString(mailEntityStrings.DescFar)); // Frontier: mail constants struct
```
Comment (VMSolidus) [inline Content.Server/Nyanotrasen/Mail/MailSystem.cs:264]:
```suggestion
                args.PushMarkup(Loc.GetString(mailEntityStrings.DescFar));
```

File: Content.Server/Nyanotrasen/Mail/MailSystem.cs [hunk 14]
```diff
@@ -241,19 +253,19 @@ private void OnAfterInteractUsing(EntityUid uid, MailComponent component, AfterI
...
                 return;
             }
 
-            args.PushMarkup(Loc.GetString("mail-desc-close", ("name", component.Recipient), ("job", component.RecipientJob)));
+            args.PushMarkup(Loc.GetString(mailEntityStrings.DescClose, ("name", component.Recipient), ("job", component.RecipientJob))); // Frontier: mail constants struct
```
Comment (VMSolidus) [inline Content.Server/Nyanotrasen/Mail/MailSystem.cs:268]:
```suggestion
            args.PushMarkup(Loc.GetString(mailEntityStrings.DescClose, ("name", component.Recipient), ("job", component.RecipientJob)));
```

File: Content.Server/Nyanotrasen/Mail/MailSystem.cs [hunk 15]
```diff
@@ -306,7 +318,17 @@ public void PenalizeStationFailedDelivery(EntityUid uid, MailComponent component
         private void OnDestruction(EntityUid uid, MailComponent component, DestructionEventArgs args)
         {
             if (component.IsLocked)
+            {
+                // DeltaV - Tampered mail recorded to logistic stats
```
Comment (VMSolidus) [inline Content.Server/Nyanotrasen/Mail/MailSystem.cs:322]:
```suggestion
```

File: Content.Server/Nyanotrasen/Mail/MailSystem.cs [hunk 16]
```diff
@@ -335,7 +357,17 @@ private void OnBreak(EntityUid uid, MailComponent component, BreakageEventArgs a
             _appearanceSystem.SetData(uid, MailVisuals.IsBroken, true);
 
             if (component.IsFragile)
+            {
+                // DeltaV - Broken mail recorded to logistic stats
```
Comment (VMSolidus) [inline Content.Server/Nyanotrasen/Mail/MailSystem.cs:361]:
```suggestion
```

File: Content.Server/Nyanotrasen/Mail/MailSystem.cs [hunk 17]
```diff
@@ -470,6 +502,15 @@ public void SetupMail(EntityUid uid, MailTeleporterComponent component, MailReci
...
             mailComp.Recipient = recipient.Name;
 
+            // Frontier: Large mail bonus
+            MailEntityStrings mailEntityStrings = mailComp.IsLarge ? MailConstants.MailLarge : MailConstants.Mail;
+            if (mailComp.IsLarge)
+            {
+                mailComp.Bounty += component.LargeBonus;
+                mailComp.Penalty += component.LargeMalus;
+            }
+            // End Frontier
```
Comment (VMSolidus) [inline Content.Server/Nyanotrasen/Mail/MailSystem.cs:512]:
```suggestion
            MailEntityStrings mailEntityStrings = mailComp.IsLarge ? MailConstants.MailLarge : MailConstants.Mail;
            if (mailComp.IsLarge)
            {
                mailComp.Bounty += component.LargeBonus;
                mailComp.Penalty += component.LargeMalus;
            }
```

File: Content.Server/Nyanotrasen/Mail/MailSystem.cs [hunk 18]
```diff
@@ -486,13 +527,24 @@ public void SetupMail(EntityUid uid, MailTeleporterComponent component, MailReci
                 mailComp.priorityCancelToken = new CancellationTokenSource();
 
                 Timer.Spawn((int) component.priorityDuration.TotalMilliseconds,
-                    () => PenalizeStationFailedDelivery(uid, mailComp, "mail-penalty-expired"),
+                    () =>
+                    {
+                        // DeltaV - Expired mail recorded to logistic stats
```
Comment (VMSolidus) [inline Content.Server/Nyanotrasen/Mail/MailSystem.cs:532]:
```suggestion
```

File: Content.Server/Nyanotrasen/Mail/MailSystem.cs [hunk 19]
```diff
@@ -486,13 +527,24 @@ public void SetupMail(EntityUid uid, MailTeleporterComponent component, MailReci
...
+
+                        PenalizeStationFailedDelivery(uid, mailComp, "mail-penalty-expired");
+                    },
                     mailComp.priorityCancelToken.Token);
             }
 
             _appearanceSystem.SetData(uid, MailVisuals.JobIcon, recipient.JobIcon);
 
-            _metaDataSystem.SetEntityName(uid, Loc.GetString("mail-item-name-addressed",
+            _metaDataSystem.SetEntityName(uid, Loc.GetString(mailEntityStrings.NameAddressed, // Frontier: move constant to MailEntityString
```
Comment (VMSolidus) [inline Content.Server/Nyanotrasen/Mail/MailSystem.cs:547]:
```suggestion
            _metaDataSystem.SetEntityName(uid, Loc.GetString(mailEntityStrings.NameAddressed,
```

File: Content.Server/Nyanotrasen/Mail/MailSystem.cs [hunk 20]
```diff
@@ -668,6 +720,8 @@ public void SpawnMail(EntityUid uid, MailTeleporterComponent? component = null)
 
                 var mail = EntityManager.SpawnEntity(chosenParcel, Transform(uid).Coordinates);
                 SetupMail(mail, component, candidate);
+
+                _tagSystem.AddTag(mail, "Mail"); // Frontier
```
Comment (VMSolidus) [inline Content.Server/Nyanotrasen/Mail/MailSystem.cs:724]:
```suggestion
                _tagSystem.AddTag(mail, "Mail");
```

File: Content.Server/Nyanotrasen/Mail/MailSystem.cs [hunk 21]
```diff
@@ -713,6 +767,21 @@ private void UpdateMailTrashState(EntityUid uid, bool isTrash)
         {
             _appearanceSystem.SetData(uid, MailVisuals.IsTrash, isTrash);
         }
+
+        // DeltaV - Helper function that executes for each StationLogisticsStatsComponent
```
Comment (VMSolidus) [inline Content.Server/Nyanotrasen/Mail/MailSystem.cs:771]:
```suggestion
        // Helper function that executes for each StationLogisticsStatsComponent
```

**Review by VMSolidus — CHANGES_REQUESTED** [review]

**Comment by VMSolidus** [comment]
Probably going to merge this after the last batch of requested changes are made.

**Comment by Mnemotechnician** [comment]
Inshalla; nynanotrasen, delta-v and frontier mail code is now merged

**Comment by VMSolidus** [comment]
> Inshalla; nynanotrasen, delta-v and frontier mail code is now merged

Fuck yea

**Review by VMSolidus — APPROVED** [review]

**Review by OldDanceJacket — APPROVED** [review]

**Review by Peptide90 — APPROVED** [review]

**Pull request merged**